### PR TITLE
[Improve][e2e] Improve end-to-end testing of connector-v2 in a multi-engine environment

### LINKED ISSUE #2733: [Improve][e2e] Improve end-to-end testing of connector-v2 in a multi-engine environment
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/incubator-seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

Now developers need to write a lot of repetitive code to test different engines, which is a very boring thing.
If we support more versions of the engine, this will generate more duplicate test code.

### Usage Scenario

Connector-v2 e2e

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ashulin**
## Purpose of this pull request
#2733

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by ashulin** [comment]
@hailin0 @CalvinKirs @Hisoka-X @EricJoy2048 @TyrantLucifer @laglangyue @liugddx @ic4y PTAL, thanks.

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/TestContainer.java [hunk 1]
```diff
@@ -33,16 +36,17 @@
 import java.util.ArrayList;
 import java.util.List;
 
-public abstract class AbstractContainer {
+public abstract class TestContainer implements TestResource {
```
Comment (EricJoy2048) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/TestContainer.java]: Modify `TestContainer ` to `AbstractTestContainer` ?

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/flink/TestFlinkContainer.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Stream;
+
+/**
+ * This class is the base class of FlinkEnvironment test.
+ * The before method will create a Flink cluster, and after method will close the Flink cluster.
+ * You can use {@link TestFlinkContainer#executeJob} to submit a seatunnel config and run a seatunnel job.
+ */
+public abstract class TestFlinkContainer extends TestContainer {
```
Comment (EricJoy2048) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/flink/TestFlinkContainer.java]: I suggest the abstract class name start with `Abstract`.

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/junit/TestCaseInvocationContextProvider.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+        int containerAmount = testContainers.size();
+        return testContainers.stream()
+            .map(testContainer -> new TestResourceProvidingInvocationContext(testContainer, containerExtendedFactory, containerAmount));
+    }
+
+    static class TestResourceProvidingInvocationContext implements TestTemplateInvocationContext {
+        private final TestContainer testContainer;
+        private final ContainerExtendedFactory containerExtendedFactory;
+
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/junit/TestCaseInvocationContextProvider.java]: remove?

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            .withEnv("SPARK_MODE", "master")
+            .withLogConsumer(new Slf4jLogConsumer(LOG))
+            .withCreateContainerCmdModifier(cmd -> cmd.withUser("root"));
+        // In most case we can just use standalone mode to execute a spark job, if we want to use cluster mode, we need to
+        // start a worker.
+        Startables.deepStart(Stream.of(master)).join();
+        copySeaTunnelStarter(master);
+        // execute extra commands
+        executeExtraCommands(master);
+        LOG.info("Spark container started");
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java]: Use `LOG.info("Spark[{}] container started", identifier());`?

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        if (master != null) {
+            master.stop();
+        }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java:68]: Add `LOG.info("Spark[{}] container closed", identifier());`?

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/flink/AbstractTestFlinkContainer.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                .withEnv("FLINK_PROPERTIES", properties)
+                .dependsOn(jobManager)
+                .withLogConsumer(new Slf4jLogConsumer(LOG));
+
+        Startables.deepStart(Stream.of(jobManager)).join();
+        Startables.deepStart(Stream.of(taskManager)).join();
+        copySeaTunnelStarter(jobManager);
+        // execute extra commands
+        executeExtraCommands(jobManager);
+        LOG.info("Flink containers are started.");
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/flink/AbstractTestFlinkContainer.java]: Use `LOG.info("Flink[{}] container are started", identifier());?`

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/flink/AbstractTestFlinkContainer.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        if (taskManager != null) {
+            taskManager.stop();
+        }
+        if (jobManager != null) {
+            jobManager.stop();
+        }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/flink/AbstractTestFlinkContainer.java:103]: Use `LOG.info("Flink[{}] container are closed", identifier());?`

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/TestSuiteBase.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+@ExtendWith({
+    ContainerTestingExtension.class,
+    TestLoggerExtension.class,
+    TestCaseInvocationContextProvider.class
+})
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public abstract class TestSuiteBase {
+
+    @TestContainers
+    TestContainersFactory containersFactory = ContainerUtil::discoverTestContainers;
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/TestSuiteBase.java]: Use `private ...`?

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java (hunk 4 above)
Comment (ashulin) [reply seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/container/spark/AbstractTestSparkContainer.java]: done

**Comment by ashulin** [comment]
@EricJoy2048 @hailin0 I modified the code  based on the code review and added the shared network. PTAL again, thanks

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.seatunnel.e2e.common.TestSuiteBase;
+import org.apache.seatunnel.e2e.common.container.TestContainer;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.TestTemplate;
+import org.testcontainers.containers.Container;
+
+import java.io.IOException;
+
+public class FakeSourceToAssertIT extends TestSuiteBase {
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java:29]: This is too simple, add a complex user case? e.g. redis

**Review by hailin0 — COMMENTED** [review]
LGTM

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java (hunk 9 above)
Comment (ashulin) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java:29]: I'll add complex case in later PR.

Comment (ashulin) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java:29]: In fact, there is only a small difference between the complex case and the current `connector-jdbc-flink-e2e` module, only the developer needs to pay attention to the data after one run of the test method may affect the next run.
The sink connector needs to be cleaned every time data or generate a specific location from `TestContainer#identifier()`.

Comment (hailin0) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-assert-e2e/src/test/java/org/apache/seatunnel/e2e/connector/assertion/FakeSourceToAssertIT.java:29]: > In fact, there is only a small difference between the complex case and the current `connector-jdbc-flink-e2e` module, only the developer needs to pay attention to the data after one run of the test method may affect the next run. The sink connector needs to be cleaned every time data or generate a specific location from `TestContainer#identifier()`.

But `connector-jdbc-flink-e2e` is not using `@TestTemplate`

**Review by EricJoy2048 — APPROVED** [review]
+1

**Review by TyrantLucifer — APPROVED** [review]

**Pull request merged**